File: daily_problems/2024/12/1230/personal_submission/cf51c_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/convolution"
// #include "atcoder/dsu"
// #include "atcoder/fenwicktree"
// #include "atcoder/lazysegtree"
// #include "atcoder/math"
// #include "atcoder/maxflow"
// #include "atcoder/mincostflow"
// #include "atcoder/modint"
// #include "atcoder/scc"
// #include "atcoder/segtree"
// #include "atcoder/string"
// #include "atcoder/twosat"
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divide, I conquer!

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n;
    cin >> n;

    vector<int> nums(n);
    for (auto &v: nums)
        cin >> v;
    sort(nums.begin(), nums.end());

    auto div2 = [&] (int x) -> string {
        if (x & 1) return to_string(x / 2) + ".500000";
        return to_string(x / 2) + ".000000";
    };

    if (n < 3) {
        cout << div2(0) << '\n';
        cout << div2(nums[0] * 2) << ' ' << div2(nums.back() * 2) << ' ' << div2(0);
    }
    else {
        int l = 0, r = n - 1;
        int cur = nums[r - 1] - nums[l + 1], idx = l + 1;

        while (l < r) {
            if (nums[l + 1] - nums[0] < nums.back() - nums[r - 1]) l ++;
            else r --;
            int v = max(nums[l] - nums[0], max(nums.back() - nums[r], nums[r - 1] - nums[l + 1]));
            if (v < cur) cur = v, idx = l + 1;
        }

        cout << div2(cur) << '\n';
        cout << div2(nums[0] * 2 + cur) << ' ' << div2(nums[idx] * 2 + cur) << ' ' << div2(nums.back() * 2 - cur);
    }

    return 0;
}

File: daily_problems/2024/12/1230/personal_submission/cf51c_yawn_sean.py
# Submission link: https://codeforces.com/contest/51/submission/299017704
def main():
    n = II()
    nums = LII()
    nums.sort()

    def div2(x):
        if x % 2: return f'{x // 2}.500000'
        return f'{x // 2}.000000'

    if n == 1 or n == 2:
        print(div2(0))
        print(div2(nums[0] * 2), div2(nums[-1] * 2), div2(0))
    else:
        l = 0
        r = n - 1
        
        cur = nums[r - 1] - nums[l + 1]
        idx = l + 1
        while l < r:
            if nums[l + 1] - nums[0] < nums[-1] - nums[r - 1]:
                l += 1
            else:
                r -= 1
            v = max(nums[l] - nums[0], nums[-1] - nums[r], nums[r - 1] - nums[l + 1])
            if v < cur: 
                cur = v
                idx = l + 1
        
        print(div2(cur))
        print(div2(nums[0] * 2 + cur), div2(nums[idx] * 2 + cur), div2(nums[-1] * 2 - cur))